subsys/net/ip/net_if.c

```diff
@@ -4360,12 +4365,23 @@ static void init_igmp(struct net_if *iface)
 
 int net_if_up(struct net_if *iface)
 {
+	const struct device *dev;
 	int status = 0;
 
 	NET_DBG("iface %p", iface);
 
 	net_if_lock(iface);
 
+	/* If the device is not ready it is pointless trying to take it up. */
+	dev = net_if_get_device(iface);
+	NET_ASSERT(dev);
+
+	if (!device_is_ready(dev)) {
```

Comment:
As per the linked bug this would fail as `dev->state->init_res` won't be set to failure in case device fails, its only set to failure if `services_init` fails.

Reply:
Hmm, could you elaborate what is the use case because you mentioned in the issue specifically the `device_is_ready()` checks?

Reply:
Coming back to this PR and I have more comments.

> Usecase1: An application that works with two network interfaces can use `device_is_ready` check to select a working interface and switch to other.

The application should not investigate the device but the network interface status if it is interested if it can send something.

> 
> Usecase2: An application that tries to bring the network interface up and implements a recovery mechanism in case if it fails multiple times, using `device_is_ready` it can skip interface up and go straight to recovery or identify that device has failed using `device_is_ready` and does chip recovery. (Of course this can be implemented in many ways, but just an example).

Here too, it is better to do things in network interface level. If the network device has implemented `start()` and `stop()`  network API functions, then the driver can return relevant status code up in the stack when calling `net_if_up()` and `net_if_down()`